File: ruby/lib/runapi/hailuo/resources/image_to_video.rb
# frozen_string_literal: true

module RunApi
  module Hailuo
    module Resources
      # Hailuo image-to-video resource.
      # Animate a still image into video guided by a text prompt and first-frame image.
      class ImageToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/hailuo/image_to_video"

        RESPONSE_CLASS = Types::VideoTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoTaskResponse

        def initialize(http)
          @http = http
        end

        # Generate an image-to-video task and wait until complete.
        #
        # @param params [Hash] image-to-video parameters
        # @return [RunApi::Hailuo::Types::CompletedVideoTaskResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an image-to-video task.
        #
        # @param params [Hash] image-to-video parameters
        # @return [RunApi::Hailuo::Types::VideoTaskResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get image-to-video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Hailuo::Types::VideoTaskResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["image-to-video"], params)

          raise Core::ValidationError, "prompt is required" unless param(params, :prompt)

          model = param(params, :model)
          duration_seconds = param(params, :duration_seconds)
          output_resolution = param(params, :output_resolution)

          case model
          when "hailuo-02-image-to-video-pro"
            raise Core::ValidationError, "duration_seconds is not supported for #{model}" if duration_seconds
            raise Core::ValidationError, "output_resolution is not supported for #{model}" if output_resolution
          when "hailuo-2.3-image-to-video-pro", "hailuo-2.3-image-to-video-standard"
            raise Core::ValidationError, "last_frame_image_url is not supported for #{model}" if param(params, :last_frame_image_url)
            raise Core::ValidationError, "prompt_optimizer is not supported for #{model}" if param(params, :prompt_optimizer)
            if duration_seconds == 10 && output_resolution.to_s == "1080p"
              raise Core::ValidationError, "1080p does not support 10-second duration"
            end
          end
        end
      end
    end
  end
end

File: python/src/runapi/hailuo/resources/image_to_video.py
"""Hailuo image-to-video resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedVideoTaskResponse,
    VideoTaskResponse,
)


class ImageToVideo(Resource):
    """Generate videos from a source image with Hailuo models."""

    ENDPOINT = "/api/v1/hailuo/image_to_video"

    RESPONSE_CLASS = VideoTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedVideoTaskResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an image-to-video task and poll until it completes.

        Args:
            **params: Image-to-video parameters (model, prompt, ...).

        Returns:
            The completed task with videos.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an image-to-video task and return immediately with an ``id``.

        Args:
            **params: Image-to-video parameters (model, prompt, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of an image-to-video task.

        Args:
            id: The task id.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["image-to-video"], params)

        if not params.get("prompt"):
            raise ValidationError("prompt is required")

        model = params.get("model")
        duration_seconds = params.get("duration_seconds")
        output_resolution = params.get("output_resolution")

        if model == "hailuo-02-image-to-video-pro":
            if duration_seconds:
                raise ValidationError(f"duration_seconds is not supported for {model}")
            if output_resolution:
                raise ValidationError(f"output_resolution is not supported for {model}")
        elif model in ("hailuo-2.3-image-to-video-pro", "hailuo-2.3-image-to-video-standard"):
            if params.get("last_frame_image_url"):
                raise ValidationError(f"last_frame_image_url is not supported for {model}")
            if params.get("prompt_optimizer"):
                raise ValidationError(f"prompt_optimizer is not supported for {model}")
            if duration_seconds == 10 and str(output_resolution) == "1080p":
                raise ValidationError("1080p does not support 10-second duration")
